thorlcr/activities/thactivityutil.cpp

```diff
@@ -692,6 +682,26 @@ class CWriteHandler : implements IFileIO, public CInterface
         if (partDesc.numCopies()>1)
             _doReplicate(&activity, partDesc, iProgress);
     }
+
+public:
+    IMPLEMENT_IINTERFACE_USING(CInterface);
+
+    CWriteHandler(CActivityBase &_activity, IPartDescriptor &_partDesc, IFile *_primary, IFileIO *_primaryio, ICopyFileProgress *_iProgress, unsigned _twFlags, bool *_aborted)
+        : activity(_activity), partDesc(_partDesc), primary(_primary), primaryio(_primaryio), iProgress(_iProgress), twFlags(_twFlags), aborted(_aborted), fipScope(primary->queryFilename())
+    {
+        RemoteFilename rfn;
+        partDesc.getFilename(0, rfn);
+        remote = !rfn.isLocal();
+        rfn.getPath(primaryName);
+        if (globals->getPropBool("@replicateAsync", true))
+            cancelReplicates(&activity, partDesc);
+    }
+    virtual void beforeDispose() override
+    {
+        // Can't throw in destructor...
+        // Note that if we do throw the CWriteHandler object is liable to be leaked...
+        checkAndHandleClose();
```

Comment:
I'm thinking about this issue for a similar reason.

I'm not sure this change is safe - e.g., if this is being deleted because the owning object is being destroyed then you will still be throwing an exception within a destructor - and will cause the query to terminate.

Reply:
I agree, with or without this change this code could cause the process to terminate.
Not quite the issue this change is making though.
It's giving a chance to directly call this code (via overridden close()) before it is indirectly called via beforeDispose/dtor.

Re. possible termination if this code is hit via dtor (because close() not called), as with ~CFileIO, it could be wrapped in a try/catch and log it and continue, but ... in some ways that's worse than terminating with a core and stack trace. You don't know the file(s) state if you continue.

If explicitly called now (as it should be) via close() - then this beforeDispose will be a nop.
@ghalliday

Reply:
On rereading I think it is an improvment on the previous code.
It is still potentially a problem throwing an exception from beforeDispose() - it will cause program termination if the Release is being called from a destructor (rather than an explicit clear()).  It would probably be safer to catch any exception and log it instead.

Reply:
This was supposed to be an improvement - I suspect there are other cases where ~CFileIO will lose exceptions on close.

It could still terminate, as you say, if dtor's via another route.
I agree, in this case, I think it would be better to catch and log and continue, because it should have already flagged the error.
I'll add that.